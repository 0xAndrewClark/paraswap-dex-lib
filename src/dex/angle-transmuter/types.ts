import { BigNumber } from 'ethers';
import { Address, Token } from '../../types';
import { ChainLinkState } from '../../lib/chainlink';

export type PoolState = {
  stablecoin: Token;
  transmuter: TransmuterState;
  oracles: {
    chainlink: { [address: string]: ChainLinkState };
    pyth: PythState;
  };
};

export type CollateralState = {
  whitelist: { status: boolean; data: string };
  fees: Fees;
  stablecoinsIssued: number;
  config: Oracle;
};

export type TransmuterState = {
  collaterals: {
    [token: string]: CollateralState;
  };
  isWhitelisted: {
    [type: number]: Set<string>;
  };
  xRedemptionCurve: number[];
  yRedemptionCurve: number[];
  totalStablecoinIssued: number;
};

export type ChainlinkConfig = {
  [address: string]: { proxy: Address; aggregator: Address; decimals: number };
};

export type PythConfig = { proxy: Address; ids: string[] };

export type PoolConfig = {
  agEUR: Token;
  transmuter: Address;
  collaterals: Address[];
  oracles: {
    chainlink: ChainlinkConfig;
    backed: ChainlinkConfig;
    redstone: ChainlinkConfig;
    pyth: PythConfig;
  };
};

export type ChainlinkState = {
  answer: number;
  timestamp: number;
};

export type PythState = {
  [address: string]: {
    answer: number;
    expo: number;
    timestamp: number;
  };
};

export type AngleTransmuterData = {
  exchange: Address;
};

export type DexParams = {
  EURA: Token;
  transmuter: Address;
  pyth: Address;
};

export enum QuoteType {
  MintExactInput = 0,
  MintExactOutput = 1,
  BurnExactInput = 2,
  BurnExactOutput = 3,
}

export enum OracleReadType {
  CHAINLINK_FEEDS = 0,
  EXTERNAL = 1,
  NO_ORACLE = 2,
  STABLE = 3,
  WSTETH = 4,
  CBETH = 5,
  RETH = 6,
  SFRXETH = 7,
  PYTH = 8,
  MAX = 9,
  MORPHO_ORACLE = 10,
}

export enum OracleQuoteType {
  UNIT = 0,
  TARGET = 1,
}

export type Fees = {
  xFeeMint: number[];
  yFeeMint: number[];
  xFeeBurn: number[];
  yFeeBurn: number[];
};

export type Pyth = {
  pyth: Address;
  feedIds: string[];
  stalePeriods: number[];
  isMultiplied: number[];
  quoteType: OracleQuoteType;
};

export type Chainlink = {
  circuitChainlink: Address[];
  stalePeriods: number[];
  circuitChainIsMultiplied: number[];
  chainlinkDecimals: number[];
  quoteType: OracleQuoteType;
};

export type OracleFeed = {
  isChainlink: boolean;
  isPyth: boolean;
  chainlink?: Chainlink;
  pyth?: Pyth;
  otherContract?: Address;
};

export type Oracle = {
  oracleType: OracleReadType;
  targetType: OracleReadType;
  externalOracle?: Address;
  oracleFeed: OracleFeed;
  targetFeed: OracleFeed;
  hyperparameters: string;
};

export type DecodedOracleConfig = {
  oracleType: OracleReadType;
  targetType: OracleReadType;
  oracleData: string;
  targetData: string;
  hyperparameters: string;
};

export type OracleHyperparameter = {
  userDeviation: BigNumber;
  burnRatioDeviation: BigNumber;
};

export type DecodedStateMultiCallResultPythOracle = {
  price: BigNumber;
  conf: number;
  expo: number;
  publishTime: number;
};

export const BASE_9 = 1; // 1e9
export const BASE_12 = 1e3; // 1e12

export const MAX_BURN_FEE = 0.999; // 999_000_000
