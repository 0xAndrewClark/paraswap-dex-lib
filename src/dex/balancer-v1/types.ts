import { Address } from '../../types';
import BigNumber from 'bignumber.js';

export interface Token {
  address: string;
  balance: BigNumber;
  decimals: number;
  denomWeight: BigNumber;
}

export interface Pool {
  id: string;
  swapFee: BigNumber;
  totalWeight: BigNumber;
  tokens: Token[];
  tokensList: string[];
}

export type PoolState = {
  [address: string]: Pool;
};

export type BalancerV1Data = {
  pool: Address;
  exchangeProxy: Address;
};

export type DexParams = {
  subgraphURL: string;
};
