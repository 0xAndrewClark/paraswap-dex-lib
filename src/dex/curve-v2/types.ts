import { Address, NumberAsString } from '@paraswap/core';

export type CurveV2DirectSwapParam = [
  NumberAsString, // packed curvedata
  number, // i
  number, // j
  Address, // pool address
  Address, // src token
  Address, // dest token,
  NumberAsString, // fromAmount
  NumberAsString, // toAmount
  NumberAsString, // quotedAmount,
  string, // metadata - bytes32
  Address, // beneficiary
];

export enum CurveV2SwapType {
  EXCHANGE,
  EXCHANGE_UNDERLYING,
  EXCHANGE_GENERIC_FACTORY_ZAP,
}
