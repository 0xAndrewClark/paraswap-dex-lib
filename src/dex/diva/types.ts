export type DivaParams = {
  divETH: string;
  wdivETH: string;
};

export type DivETHPoolState = {
  totalShares: bigint;
  totalEther: bigint;
}

