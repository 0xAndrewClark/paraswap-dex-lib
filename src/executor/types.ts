export enum Flag {
  FIFTEEN = 15,
  ELEVEN = 11,
  NINE = 9,
  EIGHT = 8,
  SEVEN = 7,
  FIVE = 5,
  FOUR = 4,
  THREE = 3,
  ZERO = 0,
}

export enum SpecialDex {
  DEFAULT = 0,
  // SWAP_ON_MAKER_PSM = 3, // swapOnMakerPSM
  SWAP_ON_SWAAP_V2_SINGLE = 1, // swapOnSwaapV2Single
  SEND_NATIVE = 4, // sendNative
  SWAP_ON_BALANCER_V2 = 5, // swapOnBalancerV2
  SWAP_ON_UNISWAP_V2_FORK = 6,
  SWAP_ON_DYSTOPIA_UNISWAP_V2_FORK = 7,
  SWAP_ON_DYSTOPIA_UNISWAP_V2_FORK_WITH_FEE = 8,
  SWAP_ON_AUGUSTUS_RFQ = 9, // swapOnAugustusRFQ
}

export enum Executors {
  ONE = 'Executor01',
  TWO = 'Executor02',
  THREE = 'Executor03',
}

export enum RouteExecutionType {
  SINGLE_STEP = 0, // simpleSwap with 100% on a path and single DEX
  HORIZONTAL_SEQUENCE = 1, // multiSwap with 100% on each path
  VERTICAL_BRANCH = 3, // simpleSwap with N DEXs on a path
  VERTICAL_BRANCH_HORIZONTAL_SEQUENCE = 4, // multiSwap, megaSwap
  // NESTED_VERTICAL_BRANCH_HORIZONTAL_SEQUENCE = 5, // megaSwap
}
