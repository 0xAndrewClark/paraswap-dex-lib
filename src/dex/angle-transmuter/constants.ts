export const CBETH = '0xBe9895146f7AF43049ca1c1AE358B0541Ea49704';
export const RETH = '0xae78736Cd615f374D3085123A210448E74Fc6393';
export const STETH = '0xae7ab96520DE3A18E5e111B5EaAb095312D7fE84';
export const SFRXETH = '0xac3E018457B222d93114458476f3E3416Abbe38F';
