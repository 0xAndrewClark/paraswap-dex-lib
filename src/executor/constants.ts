import { ethers } from 'ethers';

const {
  utils: { hexlify, hexZeroPad },
} = ethers;

export const BYTES_28_LENGTH = 28;
export const BYTES_64_LENGTH = 64;

export const ZEROS_4_BYTES = hexZeroPad(hexlify(0), 4);
export const ZEROS_12_BYTES = hexZeroPad(hexlify(0), 12);
export const ZEROS_20_BYTES = hexZeroPad(hexlify(0), 20);
export const ZEROS_28_BYTES = hexZeroPad(hexlify(0), 28);
export const ZEROS_32_BYTES = hexZeroPad(hexlify(0), 32);

export const SWAP_EXCHANGE_100_PERCENTAGE = 100;

export const ZERO_ROUTE_INDEX = 0;

export const EXECUTORS_FUNCTION_CALL_DATA_TYPES: string[] = [
  'bytes20', // address(bytes20)
  'bytes4', // calldata Size(bytes 4)
  'bytes2', // fromAmount Pos(bytes2)
  'bytes2', // destTokenPos(bytes2)
  'bytes1', // returnAmountPos(bytes1)
  'bytes1', // specialExchange (bytes2)
  'bytes2', // flag(bytes2)
  'bytes28', // zero padding (bytes28)
  'bytes', // dex calldata (bytes)
];

export const APPROVE_CALLDATA_DEST_TOKEN_POS = 68;
export const WRAP_UNWRAP_FROM_AMOUNT_POS = 4;
export const DEFAULT_RETURN_AMOUNT_POS = '0xff';
