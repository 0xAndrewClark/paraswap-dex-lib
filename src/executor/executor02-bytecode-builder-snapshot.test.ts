import dotenv from 'dotenv';
dotenv.config();

import { Network } from '../constants';
import { DummyDexHelper } from '../dex-helper';
import { Executor02BytecodeBuilder } from './Executor02BytecodeBuilder';
import { OptimalRate } from '@paraswap/core';
import { DexExchangeParam } from '../types';
import { DepositWithdrawReturn } from '../dex/weth/types';

import priceRouteSimpleSwapSushiV3UniV3SushiEth from './fixtures/executor02/routes/price-route-simpleSwap-sushiv3-univ3-sushi-eth.json';
import exchangeParamsSimpleSwapSushiV3UniV3SushiEth from './fixtures/executor02/exchange-params/price-route-simpleSwap-sushiv3-univ3-sushi-eth.json';
import maybeWethCalldataSimpleSwapSushiV3UniV3SushiEth from './fixtures/executor02/maybe-weth-calldata/price-route-simpleSwap-sushiv3-univ3-sushi-eth.json';

import priceRouteSimpleSwapSushiV3UniV3EthSushi from './fixtures/executor02/routes/price-route-simpleSwap-sushiv3-univ3-eth-sushi.json';
import exchangeParamsSimpleSwapSushiV3UniV3EthSushi from './fixtures/executor02/exchange-params/price-route-simpleSwap-sushiv3-univ3-eth-sushi.json';
import maybeWethCalldataSimpleSwapSushiV3UniV3EthSushi from './fixtures/executor02/maybe-weth-calldata/price-route-simpleSwap-sushiv3-univ3-eth-sushi.json';

import priceRouteSimpleSwapSushiV3BalancerV1EthUsdc from './fixtures/executor02/routes/price-route-simpleSwap-sushiv3-balancerv1-eth-usdc.json';
import exchangeParamsSimpleSwapSushiV3BalancerV1EthUsdc from './fixtures/executor02/exchange-params/price-route-simpleSwap-sushiv3-balancerv1-eth-usdc.json';
import maybeWethCalldataSimpleSwapSushiV3BalancerV1EthUsdc from './fixtures/executor02/maybe-weth-calldata/price-route-simpleSwap-sushiv3-balancerv1-eth-usdc.json';

import priceRouteSimpleSwapSushiV3BalancerV1UsdcEth from './fixtures/executor02/routes/price-route-simpleSwap-sushiv3-balancerv1-usdc-eth.json';
import exchangeParamsSimpleSwapSushiV3BalancerV1UsdcEth from './fixtures/executor02/exchange-params/price-route-simpleSwap-sushiv3-balancerv1-usdc-eth.json';
import maybeWethCalldataSimpleSwapSushiV3BalancerV1UsdcEth from './fixtures/executor02/maybe-weth-calldata/price-route-simpleSwap-sushiv3-balancerv1-usdc-eth.json';

import priceRouteSimpleSwapUniV3CurveV1UsdtDai from './fixtures/executor02/routes/price-route-simpleSwap-univ3-curvev1-usdt-dai.json';
import exchangeParamsSimpleSwapUniV3CurveV1UsdtDai from './fixtures/executor02/exchange-params/price-route-simpleSwap-univ3-curvev1-usdt-dai.json';

describe('Executor02BytecodeBuilder Snapshot tests', () => {
  let executor02BytecodeBuilder: Executor02BytecodeBuilder;
  beforeEach(() => {
    const network = Network.MAINNET;
    const dexHelper = new DummyDexHelper(network);
    executor02BytecodeBuilder = new Executor02BytecodeBuilder(dexHelper);
  });

  describe('buildByteCode', () => {
    describe('SimpleSwap', () => {
      it('should produce correct bytecode for simpleSwap ETH -> SUSHI via SushiSwapV3 and UniswapV3', () => {
        const bytecode = executor02BytecodeBuilder.buildByteCode(
          priceRouteSimpleSwapSushiV3UniV3SushiEth as unknown as OptimalRate,
          exchangeParamsSimpleSwapSushiV3UniV3SushiEth as unknown as DexExchangeParam[],
          maybeWethCalldataSimpleSwapSushiV3UniV3SushiEth as unknown as DepositWithdrawReturn,
        );

        expect(bytecode).toMatchSnapshot();
      });

      it('should produce correct bytecode for simpleSwap SUSHI -> ETH via SushiSwapV3 and UniswapV3', () => {
        const bytecode = executor02BytecodeBuilder.buildByteCode(
          priceRouteSimpleSwapSushiV3UniV3EthSushi as unknown as OptimalRate,
          exchangeParamsSimpleSwapSushiV3UniV3EthSushi as unknown as DexExchangeParam[],
          maybeWethCalldataSimpleSwapSushiV3UniV3EthSushi as unknown as DepositWithdrawReturn,
        );

        expect(bytecode).toMatchSnapshot();
      });

      it('should produce correct bytecode for simpleSwap ETH -> USDC via SushiSwapV3 and BalancerV1', () => {
        const bytecode = executor02BytecodeBuilder.buildByteCode(
          priceRouteSimpleSwapSushiV3BalancerV1EthUsdc as unknown as OptimalRate,
          exchangeParamsSimpleSwapSushiV3BalancerV1EthUsdc as unknown as DexExchangeParam[],
          maybeWethCalldataSimpleSwapSushiV3BalancerV1EthUsdc as unknown as DepositWithdrawReturn,
        );

        expect(bytecode).toMatchSnapshot();
      });

      it('should produce correct bytecode for simpleSwap USDC -> ETH via SushiSwapV3 and BalancerV1', () => {
        const bytecode = executor02BytecodeBuilder.buildByteCode(
          priceRouteSimpleSwapSushiV3BalancerV1UsdcEth as unknown as OptimalRate,
          exchangeParamsSimpleSwapSushiV3BalancerV1UsdcEth as unknown as DexExchangeParam[],
          maybeWethCalldataSimpleSwapSushiV3BalancerV1UsdcEth as unknown as DepositWithdrawReturn,
        );

        expect(bytecode).toMatchSnapshot();
      });

      it('should produce correct bytecode for simpleSwap USDT -> DAI via UniswapV3 and CurveV1', () => {
        const bytecode = executor02BytecodeBuilder.buildByteCode(
          priceRouteSimpleSwapUniV3CurveV1UsdtDai as unknown as OptimalRate,
          exchangeParamsSimpleSwapUniV3CurveV1UsdtDai as unknown as DexExchangeParam[],
          undefined, // no weth calldata
        );

        expect(bytecode).toMatchSnapshot();
      });
    });
  });
});
