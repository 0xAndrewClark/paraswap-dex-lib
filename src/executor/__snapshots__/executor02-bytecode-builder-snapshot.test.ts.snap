// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Executor02BytecodeBuilder Snapshot tests buildByteCode SimpleSwap should produce correct bytecode for simpleSwap ETH -> SUSHI via SushiSwapV3 and UniswapV3 1`] = `"0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000004a00000000000000000000000000000004000000000000000000000000000002710c02aaa39b223fe8d0a0e5c4f27ead9083c756cc200000020000400000000000900000000000000000000000000000000000000000000000000000000d0e30db0000000000000000000000000000001e000000000000000000000000000001450c02aaa39b223fe8d0a0e5c4f27ead9083c756cc200000060000000440000000000000000000000000000000000000000000000000000000000000000095ea7b3000000000000000000000000e592427a0aece92de3edee1f18e0157c058615640000000000000000000000000000000000000000000000001cdda4faccd00000e592427a0aece92de3edee1f18e0157c0586156400000140000000000000000000000000000000000000000000000000000000000000000000000000c04b8d59000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000e6f1c1f58e34df3892cb1b1c193f82462bc0a86e0000000000000000000000000000000000000000000000000000000065898ab80000000000000000000000000000000000000000000000001cdda4faccd000000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000002bc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000bb86b3595068778dd592e39a122f4f5a5cf09c90fe2000000000000000000000000000000000000000000000000000000000000000000000001e0000000000000000000000000000012c0c02aaa39b223fe8d0a0e5c4f27ead9083c756cc200000060000000440000000000000000000000000000000000000000000000000000000000000000095ea7b300000000000000000000000000f23572b16c5e9e58e7b965def51ff8ff546e340000000000000000000000000000000000000000000000001aa535d3d0c0000000f23572b16c5e9e58e7b965def51ff8ff546e3400000140000000000000000000000000000000000000000000000000000000000000000000000000c04b8d59000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000e6f1c1f58e34df3892cb1b1c193f82462bc0a86e0000000000000000000000000000000000000000000000000000000065898ab80000000000000000000000000000000000000000000000001aa535d3d0c000000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000002bc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000bb86b3595068778dd592e39a122f4f5a5cf09c90fe2000000000000000000000000000000000000000000"`;

exports[`Executor02BytecodeBuilder Snapshot tests buildByteCode SimpleSwap should produce correct bytecode for simpleSwap ETH -> USDC via SushiSwapV3 and BalancerV1 1`] = `"0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000004000000000000000000000000000001518c02aaa39b223fe8d0a0e5c4f27ead9083c756cc200000020000400000000000900000000000000000000000000000000000000000000000000000000d0e30db0000000000000000000000000000001e000000000000000000000000000000c80c02aaa39b223fe8d0a0e5c4f27ead9083c756cc200000060000000440000000000000000000000000000000000000000000000000000000000000000095ea7b300000000000000000000000000f23572b16c5e9e58e7b965def51ff8ff546e340000000000000000000000000000000000000000000000002c68af0bb140000000f23572b16c5e9e58e7b965def51ff8ff546e3400000140000000000000000000000000000000000000000000000000000000000000000000000000c04b8d59000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000e6f1c1f58e34df3892cb1b1c193f82462bc0a86e0000000000000000000000000000000000000000000000000000000065898bb00000000000000000000000000000000000000000000000002c68af0bb14000000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000002bc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000bb8a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48000000000000000000000000000000000000000000000000000000000000000000000001e000000000000000000000000000000898c02aaa39b223fe8d0a0e5c4f27ead9083c756cc200000060000000440000000000000000000000000000000000000000000000000000000000000000095ea7b300000000000000000000000000f23572b16c5e9e58e7b965def51ff8ff546e340000000000000000000000000000000000000000000000001e87f85809dc000000f23572b16c5e9e58e7b965def51ff8ff546e3400000140000000000000000000000000000000000000000000000000000000000000000000000000c04b8d59000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000e6f1c1f58e34df3892cb1b1c193f82462bc0a86e0000000000000000000000000000000000000000000000000000000065898bb00000000000000000000000000000000000000000000000001e87f85809dc00000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000002bc02aaa39b223fe8d0a0e5c4f27ead9083c756cc20001f4a0b86991c6218b36c1d19d4a2e9eb0ce3606eb4800000000000000000000000000000000000000000000000000000000000000000000000340000000000000000000000000000011f86317c5e82a06e1d8bf200d21f4510ac2c038ac81000002a00000002400000005000000000000000000000000000000000000000000000000000000004e7ffa040000000000000000000000000000000000000000000000000000000000000060000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb480000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000400000000000000000000000069d460e01070a7ba1bc363885bc8f4f0daa19bf500000000000000000000000000000000000000000000000029a2241af62c00000000000000000000000000000000000000000000000000000000000000000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0000000000000000000000001373e57f764a7944bdd7a4bd5ca3007d496934da000000000000000000000000000000000000000000000000058d15e1762800000000000000000000000000000000000000000000000000000000000000000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0000000000000000000000002257aaac34bcb27900291f7b84ee2565a6cbac570000000000000000000000000000000000000000000000000853a0d2313c00000000000000000000000000000000000000000000000000000000000000000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0000000000000000000000008a649274e4d777ffc6851f13d23a86bbfa2f2fbf0000000000000000000000000000000000000000000000000853a0d2313c00000000000000000000000000000000000000000000000000000000000000000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffa0b86991c6218b36c1d19d4a2e9eb0ce3606eb4800000060002400000000000300000000000000000000000000000000000000000000000000000000a9059cbb000000000000000000000000e6f1c1f58e34df3892cb1b1c193f82462bc0a86e00000000000000000000000000000000000000000000000000000004b97af117"`;

exports[`Executor02BytecodeBuilder Snapshot tests buildByteCode SimpleSwap should produce correct bytecode for simpleSwap SUSHI -> ETH via SushiSwapV3 and UniswapV3 1`] = `"0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000002e000000000000000000000000000000280000000000000000000000000000027106b3595068778dd592e39a122f4f5a5cf09c90fe200000060000000440000000000000000000000000000000000000000000000000000000000000000095ea7b300000000000000000000000000f23572b16c5e9e58e7b965def51ff8ff546e340000000000000000000000000000000000000000000000a2a15d09519be0000000f23572b16c5e9e58e7b965def51ff8ff546e3400000140000000ef0000000800000000000000000000000000000000000000000000000000000000c04b8d59000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000008edf3289c84e5ba9c73350805a7d4c1f43b41e080000000000000000000000000000000000000000000000000000000065898abd0000000000000000000000000000000000000000000000a2a15d09519be000000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000002b6b3595068778dd592e39a122f4f5a5cf09c90fe2000bb8c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000000000000000000000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000000400004000000000007000000000000000000000000000000000000000000000000000000002e1a7d4d000000000000000000000000000000000000000000000000144aa80473e3b150e6f1c1f58e34df3892cb1b1c193f82462bc0a86e0000002000000000000400090000000000000000000000000000000000000000000000000000000000000000"`;

exports[`Executor02BytecodeBuilder Snapshot tests buildByteCode SimpleSwap should produce correct bytecode for simpleSwap USDC -> ETH via SushiSwapV3 and BalancerV1 1`] = `"0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000008c00000000000000000000000000000024000000000000000000000000000000c80a0b86991c6218b36c1d19d4a2e9eb0ce3606eb4800000060000000440000000000000000000000000000000000000000000000000000000000000000095ea7b300000000000000000000000000f23572b16c5e9e58e7b965def51ff8ff546e34000000000000000000000000000000000000000000000000000000017d78400000f23572b16c5e9e58e7b965def51ff8ff546e3400000140000000ef0000000800000000000000000000000000000000000000000000000000000000c04b8d59000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000091e96a80943d06085997bbf41f253ae85f560b3e0000000000000000000000000000000000000000000000000000000065898c61000000000000000000000000000000000000000000000000000000017d7840000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000002ba0b86991c6218b36c1d19d4a2e9eb0ce3606eb48000bb8c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000000000000000000000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000000400004000000000007000000000000000000000000000000000000000000000000000000002e1a7d4d00000000000000000000000000000000000000000000000044a8f01fd4911f4f0000000000000000000000000000024000000000000000000000000000000960a0b86991c6218b36c1d19d4a2e9eb0ce3606eb4800000060000000440000000000000000000000000000000000000000000000000000000000000000095ea7b300000000000000000000000000f23572b16c5e9e58e7b965def51ff8ff546e34000000000000000000000000000000000000000000000000000000011e1a300000f23572b16c5e9e58e7b965def51ff8ff546e3400000140000000ef0000000800000000000000000000000000000000000000000000000000000000c04b8d59000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000091e96a80943d06085997bbf41f253ae85f560b3e0000000000000000000000000000000000000000000000000000000065898c61000000000000000000000000000000000000000000000000000000011e1a30000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000002ba0b86991c6218b36c1d19d4a2e9eb0ce3606eb480001f4c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000000000000000000000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000000400004000000000007000000000000000000000000000000000000000000000000000000002e1a7d4d00000000000000000000000000000000000000000000000044a8f01fd4911f4f000000000000000000000000000003a000000000000000000000000000001130a0b86991c6218b36c1d19d4a2e9eb0ce3606eb4800000060000000440000000000000000000000000000000000000000000000000000000000000000095ea7b30000000000000000000000006317c5e82a06e1d8bf200d21f4510ac2c038ac81000000000000000000000000000000000000000000000000000000020c8558006317c5e82a06e1d8bf200d21f4510ac2c038ac81000002c00000000000000004000000000000000000000000000000000000000000000000000000003ca9184e0000000000000000000000000000000000000000000000000000000000000080000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48000000000000000000000000000000000000000000000000000000020c8558000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000400000000000000000000000069d460e01070a7ba1bc363885bc8f4f0daa19bf5000000000000000000000000000000000000000000000000000000014dc938000000000000000000000000000000000000000000000000000000000000000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0000000000000000000000002257aaac34bcb27900291f7b84ee2565a6cbac570000000000000000000000000000000000000000000000000000000047868c000000000000000000000000000000000000000000000000000000000000000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0000000000000000000000008a649274e4d777ffc6851f13d23a86bbfa2f2fbf0000000000000000000000000000000000000000000000000000000047868c000000000000000000000000000000000000000000000000000000000000000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0000000000000000000000009b208194acc0a8ccb2a8dcafeacfbb7dcc093f81000000000000000000000000000000000000000000000000000000002faf08000000000000000000000000000000000000000000000000000000000000000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe6f1c1f58e34df3892cb1b1c193f82462bc0a86e0000002000000000000400090000000000000000000000000000000000000000000000000000000000000000"`;

exports[`Executor02BytecodeBuilder Snapshot tests buildByteCode SimpleSwap should produce correct bytecode for simpleSwap USDT -> DAI via UniswapV3 and CurveV1 1`] = `"0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000670000000000000000000000000000001f0000000000000000000000000000000c8dac17f958d2ee523a2206206994597c13d831ec700000060000000440000000000000000000000000000000000000000000000000000000000000000095ea7b3000000000000000000000000a5407eae9ba41422680e2e00537571bcc53efbfd0000000000000000000000000000000000000000000000000000002e91866680a5407eae9ba41422680e2e00537571bcc53efbfd000000d0000000940000000800000000000000000000000000000000000000000000000000000000a6417ed6000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002e918666800000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000006b175474e89094c44da98b954eedeac495271d0f6b175474e89094c44da98b954eedeac495271d0f00000060002400000000000300000000000000000000000000000000000000000000000000000000a9059cbb000000000000000000000000e6f1c1f58e34df3892cb1b1c193f82462bc0a86e0000000000000000000000000000000000000000000842c5e210143d4ff0db94000000000000000000000000000001f000000000000000000000000000002260dac17f958d2ee523a2206206994597c13d831ec700000060000000440000000000000000000000000000000000000000000000000000000000000000095ea7b3000000000000000000000000bebc44782c7db0a1a60cb6fe97d0b483032ff1c70000000000000000000000000000000000000000000000000000080103199e00bebc44782c7db0a1a60cb6fe97d0b483032ff1c7000000d00000009400000008000000000000000000000000000000000000000000000000000000003df02124000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000080103199e000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000006b175474e89094c44da98b954eedeac495271d0f6b175474e89094c44da98b954eedeac495271d0f00000060002400000000000300000000000000000000000000000000000000000000000000000000a9059cbb000000000000000000000000e6f1c1f58e34df3892cb1b1c193f82462bc0a86e0000000000000000000000000000000000000000000842c5e210143d4ff0db94000000000000000000000000000001f0000000000000000000000000000003e8dac17f958d2ee523a2206206994597c13d831ec700000060000000440000000000000000000000000000000000000000000000000000000000000000095ea7b3000000000000000000000000debf20617708857ebe4f679508e7b7863a8a8eee000000000000000000000000000000000000000000000000000000e8d7a00080debf20617708857ebe4f679508e7b7863a8a8eee000000d0000000940000000800000000000000000000000000000000000000000000000000000000a6417ed600000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e8d7a000800000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000006b175474e89094c44da98b954eedeac495271d0f6b175474e89094c44da98b954eedeac495271d0f00000060002400000000000300000000000000000000000000000000000000000000000000000000a9059cbb000000000000000000000000e6f1c1f58e34df3892cb1b1c193f82462bc0a86e0000000000000000000000000000000000000000000842c5e210143d4ff0db94"`;
